### ISSUE TITLE
[Closed] GPG key rotation

### ISSUE THREAD

**Issue by kmittman** [NVIDIA/cuda-repo-management#1]
## Mega thread for reporting metdata issues with the CUDA repositories

Is the CDN stale? Have you seen something like:
```shell
E: Failed to fetch *.deb  Hash Sum mismatch
   Hashes of expected file:
    - SHA512:$sha512, SHA256:$sha256, SHA1:$sha1 [weak], MD5Sum:$md5 [weak], Filesize:$bytes [weak]
   Hashes of received file:
    - SHA512:$sha512, SHA256:$sha256, SHA1:$sha1 [weak], MD5Sum:$md5 [weak], Filesize:$bytes [weak]
   Last modification reported: $(date -R --utc)
```

or

```shell
*.rpm: Downloading successful, but checksum doesn't match. 
Calculated: $sha256(sha256)  
Expected: $sha256(sha256)
```

### Please provide the following information in your comment:

1. The error message and the last command(s) run.

2. When was the `Release` (Debian) or `repomd.xml` (RPM) file last modified ?
   ```shell
   $ curl -I https://developer.download.nvidia.com/compute/cuda/repos/$distro/$arch/Release
   $ curl -I https://developer.download.nvidia.com/compute/cuda/repos/$distro/$arch/repodata/repomd.xml
   ```

3. The Linux distro and architecture. If cross-compiling or containerized, please mention that.
   ```shell
   $ cat /etc/os-release
   $ uname -a
   ```

4. Which NVIDIA repositories do you have enabled ?
    Do your `.list` / `.repo` files contain URLs using HTTP (port 80) or HTTPS (port 443) ?

5. Which geographic region is the machine located in ?

6. Which CDN edge node are you hitting ?

7.  Any other relevant environmental conditions (i.e. a specific Docker container image) ?

**Comment by erikwijmans** [comment]
Hi, we are getting the following error:
```
> sudo apt update
Err:1 http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64  InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4B469963BF863CC
```

Details:
1. See above for command and error
2. `Release`  was "Last-Modified: Mon, 25 Apr 2022 20:50:31 GMT"
3. 
```
> cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.6 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.6 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
> uname -a 
Linux droid 4.15.0-66-generic #75~16.04.1-Ubuntu SMP Tue Oct 1 14:01:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```
4. We have the https://developer.download.nvidia.com/compute/cuda/repos/ enabled. Using http (also tried https, same error)
5. The machine is located in Georgia, USA
6. I am not sure how to figure this out. Happy to run a command if you give me one tho.
7. None.

**Comment by kmittman** [comment]
Hi @erikwijmans 
We are in the process of rotating our GPG public keys, there will be an announcement on the NVIDIA Developer blog very soon. Currently this has rolled out only for our `ubuntu1604/x86_64` and `fedora32/x86_64` repos, the rest will follow in a bit.

The new GPG keys for the CUDA repository
* Debian-based distros: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/3bf863cc.pub
* RPM-based distros: https://developer.download.nvidia.com/compute/cuda/repos/fedora32/x86_64/D42D0685.pub

Please remove the old `7fa2af80` key
* Debian-based distros:
```sudo apt-key del 7fa2af80```

* RPM-based distros:
```sudo rpm --erase gpg-pubkey-7fa2af80*```

And enroll the new signing key (a `cuda-keyring` package is also provided)
* Debian-based distros:

```sudo apt-key adv --fetch-keys http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/3bf863cc.pub```

* RPM-based distros: please update `.repo` with package manager specific instructions (see blog post)

**Comment by erikwijmans** [comment]
Thank you for the fast reply! That fixes it.

**Comment by rgov** [comment]
I have updated my Ubuntu 20.04 to use the new key, but am hitting this issue [which has](https://github.com/NVIDIA/nvidia-docker/issues/1395) [happened over](https://gitlab.com/nvidia/container-images/cuda/-/issues/96) [the years](https://gitlab.com/nvidia/container-images/cuda/-/issues/129). 

```
Err:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64  Packages
  File has unexpected size (631054 != 481481). Mirror sync in progress? [IP: 152.195.19.142 443]
  Hashes of expected file:
   - Filesize:481481 [weak]
   - SHA256:8556d67c6d380c957f05057f448d994584a135d7ed75e5ae6bb25c3fc1070b0b
   - SHA1:c5ea9556407a3b5daec4aac530cd038e9b490441 [weak]
   - MD5Sum:a5513131dbd2d4e50f185422ebb43ac9 [weak]
  Release file created at: Mon, 25 Apr 2022 23:27:19 +0000
```

**Comment by kkoehncke** [comment]
Whilst trying to run `apt-get update` within an Ubuntu 20.04 Docker image (`nvidia/cuda:11.3.0-base-ubuntu20.04`) after following the above instructions to update the signing key, I get the following error:
```W: GPG error: https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64  Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F60F4B3D7FA2AF80
E: The repository 'https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64  Release' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```

1. The error message and the last command(s) run.
```
(base) root@75d6db735dfe:/# apt-key del 7fa2af80
OK
(base) root@75d6db735dfe:/# apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub
Executing: /tmp/apt-key-gpghome.rPUENkVVgJ/gpg.1.sh --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub
gpg: requesting key from 'https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub'
gpg: key A4B469963BF863CC: public key "cudatools <cudatools@nvidia.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
(base) root@75d6db735dfe:/# apt update
...
W: GPG error: https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64  Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F60F4B3D7FA2AF80
E: The repository 'https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64  Release' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```

When was the Release (Debian) or repomd.xml (RPM) file last modified ?
```
(base) root@75d6db735dfe:/# curl -I https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/Release
HTTP/2 200
accept-ranges: bytes
age: 140
cache-control: max-age=604800
content-type: application/octet-stream
date: Thu, 28 Apr 2022 20:22:02 GMT
etag: "3751684597"
expires: Thu, 05 May 2022 20:22:02 GMT
last-modified: Mon, 25 Apr 2022 23:27:19 GMT
server: ECAcc (lab/772A)
x-cache: HIT
x-vdms-version: 2.9.4
content-length: 690
```
```
(base) root@75d6db735dfe:/# curl -I https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/repodata/repomd.xml
HTTP/2 404
cache-control: max-age=604800
content-type: text/html
date: Thu, 28 Apr 2022 20:24:40 GMT
expires: Thu, 05 May 2022 20:24:40 GMT
server: EOS (vny/1B6D)
x-vdms-version: 2.9.4
content-length: 445
```
3. The Linux distro and architecture. If cross-compiling or containerized, please mention that.
```
(base) root@75d6db735dfe:/# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```
```
(base) root@75d6db735dfe:/# uname -a
Linux 75d6db735dfe 5.4.0-107-generic #121-Ubuntu SMP Thu Mar 24 16:04:27 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

4. Which NVIDIA repositories do you have enabled ? 

Not sure, definitely CUDA and compute/machine- learning

5. Do your .list / .repo files contain URLs using HTTP (port 80) or HTTPS (port 443) ?

Both.

6. Which geographic region is the machine located in ?

Seattle, Washington

7. Which CDN edge node are you hitting ?

Not sure.

8. Any other relevant environmental conditions (i.e. a specific Docker container image) ?

Using the `nvidia/cuda:11.3.0-base-ubuntu20.04` official image provided by NVIDIA

**Comment by rgov** [comment]
@kkoehncke I think you may have both `cuda.list` (older) and `cuda-ubuntu2004-x86_64.list` (newer) in `/etc/apt/sources.list.d`. Just delete the older one.

@kmittman I would suggest using individual issues instead of using a megathread, since GitHub issues are not threaded and it will be very confusing dealing with several unrelated problems. And everyone who posts a question here will be subscribed to future notifications to unrelated problems which is a little annoying.

**Comment by kmittman** [comment]
Hi @topinfrassi01 
I believe that container image is maintained by Google ? Now that our repositories are updated, the next step is to roll out updated CUDA base images, and notify downstream maintainers to re-base.

**Comment by kmittman** [comment]
Hi @kkoehncke please `grep` for `developer.download.nvidia.com` in your `/etc/apt/sources.list` if you had previously use `add-apt-repository` it gets appended there, rather than a separate `.list` file in `/etc/apt/sources.list.d/`.
You'll need to search both locations and remove any existing entries.

**Comment by kmittman** [comment]
Hi @rgov
Yes, we've been hitting CDN issues. Unfortunately it took a very long time (and a few retries for posting and purge) to propagate to the edge nodes.

RE: mega thread, the problem is I cannot figure out how to configure it to send me email notifications, as technically this repo is owned by the _NVIDIA_ user/group. Is there some way to create an issue template?

**Comment by rgov** [comment]
@kmittman Yep there are [issue templates](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository) and you can use the "Watch" button (next to "Fork" in the top right) to get notified of all new issues.

**Comment by pablete** [comment]
Lots of nvidia documentation / repositories need to be changed now.... 

For example this guide on how to use the latest TensorRT 
https://docs.nvidia.com/deeplearning/tensorrt/quick-start-guide/index.html

has the the instruction to add the 7fa2af80.pub explicitly
`sudo apt-key add /var/nv-tensorrt-repo-${os}-${tag}/7fa2af80.pub`

**Comment by ericmclachlan** [comment]
> Hi @erikwijmans We are in the process of rotating our GPG public keys, there will be an announcement on the NVIDIA Developer blog very soon. Currently this has rolled out only for our `ubuntu1604/x86_64` and `fedora32/x86_64` repos, the rest will follow in a bit.
> 
> The new GPG keys for the CUDA repository
> 
> * Debian-based distros: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/3bf863cc.pub
> * RPM-based distros: https://developer.download.nvidia.com/compute/cuda/repos/fedora32/x86_64/D42D0685.pub
> 
> Please remove the old `7fa2af80` key
> 
> * Debian-based distros:
>   `sudo apt-key del 7fa2af80`
> * RPM-based distros:
>   `sudo rpm --erase gpg-pubkey-7fa2af80*`
> 
> And enroll the new signing key (a `cuda-keyring` package is also provided)
> 
> * Debian-based distros:
> 
> `sudo apt-key adv --fetch-keys http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/3bf863cc.pub`
> 
> * RPM-based distros: please update `.repo` with package manager specific instructions (see blog post)

Is there a best practice for making applying this change to nvidia docker images?

For instance, I'm using `11.4.2-cudnn8-runtime-ubuntu20.04` and I have confirmed that I'm now able to run `apt update` after running the following commands on the container:
```sh
apt-key del 7fa2af80
apt-key adv --fetch-keys http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/3bf863cc.pub
```

It seems then that running the above commands from the Dockerfile that deploys the image may be a good way of avoiding the NO_PUBKEY GPG error.

```Dockerfile
RUN apt-key del 7fa2af80
RUN apt-key adv --fetch-keys http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/3bf863cc.pub
```

TL;DR: I'm not sure how often GPG keys are rotated and I was wondering whether the above solution is a good workaround or actually a long-term solution to the problem? And if not, is there a good way to apply this fix to published docker images?

**Comment by LukeAI** [comment]
Is the key service is down?
```
wget https://developer.download.nvidia.com/compute/cuda/repos/$distro/$arch/cuda-keyring_1.0-1_all.deb
--2022-06-21 13:04:04--  https://developer.download.nvidia.com/compute/cuda/repos///cuda-keyring_1.0-1_all.deb
Resolving developer.download.nvidia.com (developer.download.nvidia.com)... 152.199.20.126
Connecting to developer.download.nvidia.com (developer.download.nvidia.com)|152.199.20.126|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2022-06-21 13:04:05 ERROR 404: Not Found.
```

**Comment by mrgzg1** [comment]
Seeing this happen on the west coast

```~~~$ curl -I https://developer.download.nvidia.com/compute/cuda/repos/ubuntu_1804/x86_64/Release
HTTP/2 404
cache-control: max-age=604800
content-type: text/html
date: Fri, 15 Jul 2022 23:44:04 GMT
expires: Fri, 22 Jul 2022 23:44:04 GMT
server: EOS (vny/0451)
content-length: 445

~~~$ curl -I https://developer.download.nvidia.com/compute/cuda/repos/ubuntu_1804/x86_64/repodata/repomd.xml
HTTP/2 404
cache-control: max-age=604800
content-type: text/html
date: Fri, 15 Jul 2022 23:44:18 GMT
expires: Fri, 22 Jul 2022 23:44:18 GMT
server: EOS (vny/0452)
content-length: 445

~~~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.6 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.6 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
mgandhi@capacity-ewr-00-station-00-04:/var/lib/dpkg/info$ uname -a
Linux capacity-ewr-00-station-00-04 4.15.0-189-generic #200-Ubuntu SMP Wed Jun 22 19:53:37 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux|
```

4. HTTPS: 
`https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/`

5. US-West

6. Not sure?

**Comment by kmittman** [comment]
Closing this again. If experience further errors, please [file a new issue](https://github.com/NVIDIA/cuda-repo-management/issues/new/choose), thank you.
